### PR TITLE
Bugfix/slot super

### PR DESCRIPTION
**Pull request by Tinche**
May $DEITY have mercy on our souls.

### CODE REVIEW

**Comment by Tinche** [comment]
Hm, I think this closure cell rewriting stuff is actually Python 3-only (the PEP that started it all seems to be https://www.python.org/dev/peps/pep-3135/, which is a Py3 PEP). So our PyPy implementation is currently superfluous since we don't test on any PyPy3 versions.

So I should either remove the PyPy-specific implementation or add a recent PyPy3 version to our .travis.yml. I think the latter?

**Comment by hynek** [comment]
Pretty sure there’s can’t be a $DEITY after this PR, so don’t worry your pretty head about it.

What we’ll have to worry about is this tho:

2017-08-02 at 09 24

That’s totally unacceptable so let’s run it normally and put a pragma no cover with an explanation next to it.

File: src/attr/_compat.py [hunk 1]
```diff
@@ -1,10 +1,12 @@
 from __future__ import absolute_import, division, print_function
 
+import platform
 import sys
 import types
 
 
 PY2 = sys.version_info[0] == 2
+PYPY = platform.python_implementation() == 'PyPy'
```
Comment (hynek) [inline src/attr/_compat.py]: ehm "

File: src/attr/_compat.py [hunk 2]
```diff
@@ -88,3 +90,12 @@ def iterkeys(d):
 
     def metadata_proxy(d):
         return types.MappingProxyType(dict(d))
+
+if PYPY:
```
Comment (hynek) [inline src/attr/_compat.py]: yeah pragma no cover and be done with it

File: src/attr/_make.py [hunk 3]
```diff
@@ -373,11 +380,19 @@ def wrap(cls):
                 # It might not actually be in there, e.g. if using 'these'.
                 cls_dict.pop(ca_name, None)
             cls_dict.pop("__dict__", None)
+            old_cls = cls
 
             qualname = getattr(cls, "__qualname__", None)
             cls = type(cls)(cls.__name__, cls.__bases__, cls_dict)
             if qualname is not None:
                 cls.__qualname__ = qualname
+            for item in cls.__dict__.values():
```
Comment (hynek) [inline src/attr/_make.py]: Dude, we *need* to explain this heresy in a comment. :D

File: src/attr/_make.py [hunk 4]
```diff
@@ -373,11 +380,19 @@ def wrap(cls):
...
+                closure_cells = getattr(item, '__closure__', [])
```
Comment (hynek) [inline src/attr/_make.py]: Y U HATE PRETTY QUOTES

File: tests/test_funcs.py [hunk 5]
```diff
@@ -154,6 +154,7 @@ def test_roundtrip(self, cls, dict_class):
         assert instance == roundtrip_instance
 
     @given(simple_classes())
+    @settings(suppress_health_check=[HealthCheck.too_slow])
```
Comment (hynek) [inline tests/test_funcs.py]: this can be removed once it's out of coverage

File: tests/test_make.py [hunk 6]
```diff
@@ -697,6 +697,7 @@ class TestMetadata(object):
     Tests for metadata handling.
     """
     @given(sorted_lists_of_attrs)
+    @settings(suppress_health_check=[HealthCheck.too_slow])
```
Comment (hynek) [inline tests/test_make.py]: same

File: tests/test_make.py [hunk 7]
```diff
@@ -718,6 +719,7 @@ def test_metadata_present(self, list_of_attrs):
                             class_attr.metadata.get(k))
 
     @given(simple_classes(), text())
+    @settings(suppress_health_check=[HealthCheck.too_slow])
```
Comment (hynek) [inline tests/test_make.py]: same

File: tests/test_make.py [hunk 8]
```diff
@@ -743,6 +745,7 @@ def test_metadata_immutability(self, C, string):
                     a.metadata.popitem()
 
     @given(lists(simple_attrs_without_metadata, min_size=2, max_size=5))
+    @settings(suppress_health_check=[HealthCheck.too_slow])
```
Comment (hynek) [inline tests/test_make.py]: same

File: tests/test_slots.py [hunk 9]
```diff
@@ -298,3 +316,52 @@ class C2(C1Bare):
     hash(c2)  # Just to assert it doesn't raise.
 
     assert {"x": 1, "y": 2, "z": "test"} == attr.asdict(c2)
+
+
+@pytest.mark.skipif(PY2, reason="closure cell rewriting is PY3-only.")
+def test_closure_cell_rewriting():
```
Comment (hynek) [inline tests/test_slots.py]: Optional: maybe parametrize?

File: tox.ini [hunk 10]
```diff
@@ -20,6 +20,10 @@ commands = coverage run --parallel -m pytest {posargs}
 deps = -rdev-requirements.txt
 commands = coverage run --parallel -m pytest {posargs}
 
+[testenv:pypy3]
```
Comment (hynek) [inline tox.ini]: 🔥

File: src/attr/_make.py [hunk 11]
```diff
@@ -373,11 +380,19 @@ def wrap(cls):
...
+                if not closure_cells:
```
Comment (hynek) [inline src/attr/_make.py]: Please use a specific one – any reason you use [] and not None in 390 anyways?  I guess you could remove the whole if statement?

**Comment by hynek** [comment]
(Also consider yourself scolded for writing code at 2am in the morning. ಠ_ಠ 🤗)

File: tests/test_slots.py (hunk 9 above)
Comment (Tinche) [inline tests/test_slots.py]: Can you clarify this comment please?

Comment (hynek) [inline tests/test_slots.py]: It should be possible to use `@pytest.mark.parametrize`?

Comment (Tinche) [inline tests/test_slots.py]: Hm is this a pytest feature I don't completely understand? I usually use parametrize to provide multiple values of a fixture.

**Comment by Tinche** [comment]
Think I got them all except the parametrization comment. I'm leaving tomorrow on a 10 day trip to Scotland (😀) so you'll either have to wait until I return or drag it over the finish line yourself.

**Pull request merged**